custom_dataset.py

```diff
@@ -0,0 +1,408 @@
+            boxes = np.zeros((len(scales), 4))
+            for i, scale in enumerate(scales):
+                x1 = y1 = 0.5 - (0.5 * scale)
+                x2 = y2 = 0.5 + (0.5 * scale)
+                boxes[i] = [x1, y1, x2, y2]
+            crops = tf.image.crop_and_resize([img], boxes=boxes,
+                                             box_indices=np.zeros(len(scales)),
+                                             crop_size=(self._HEIGHT,
+                                                        self._WIDTH))
+            # I am personally shamed of this implementation here
+            # TODO: Change distribution here
+            # TODO: Add fault proof
+            idx = tf.random.truncated_normal(shape=[], stddev=self._ZOOM_STDDEV)
+            idx = tf.math.abs(idx) # idx >= 0
+            idx = tf.math.round(idx) # Bankers rounding again
+            idx = tf.cast(idx, tf.dtypes.int32)
+            img = crops[idx]
+        return img, label
+
+    def _resize(self, img, label):
+        """Resize image for compatibility with Keras model.
+        TODO: Add custom CNN models to avoid resizing
+
+        Args:
+            img: tf.Tensor: image tensor in type tf.float32.
+            label: tf.Tensor: label tensor.
+
+        Returns:
+            img: tf.Tensor: image tensor in type tf.float32.
+            label: tf.Tensor: label tensor.
+        """
+        if self._RESIZE:
+            img = tf.image.resize(img, (self._RESIZE_HEIGHT,
+                                        self._RESIZE_WIDTH))
+        return img, label
+
+    def get_train_dataset(self, filter_size=None):
+        """Get training dataset. For the purpose of triplet selection, restrict
+        dataset to have certain number of labels if specified. See
+        https://www.tensorflow.org/addons/tutorials/losses_triplet.
+
+        Args:
+            filter_size: Int or None, if filter_size is None, do nothing.
+                         If filter_size is Int, restrict datset to only have
+                         filter_size number of classes. The classes to include
+                         is randomly selected form all classes.
+
+        Returns:
+            ds: tf.Dataset, TensorFlow dataset object for training. Each entry
+                is (image, label) pair.
+        """
+        # Get filename dataset (each entry is a filename)
+        data_dir = pathlib.Path(self._TRAIN_DATA_DIR)
+        list_ds = tf.data.Dataset.list_files(str(data_dir / '*'))
+
+        # Get labeled dataset (each entry is (image, label) tuple)
+        ds = list_ds.map(self._process_path, num_parallel_calls=AUTOTUNE)
+
+        if filter_size:
+            # Filter using filter_size
+            labels = tf.constant(np.random.choice(1000, filter_size,
+                                                  replace=False))
+            ds = ds.filter(lambda img, label:
+                           tf.reduce_any(tf.equal(label,labels)))
+
+        # Format conversion
```

Comment:
In stead of write the code as 
```
# Do X
ds = ds.x()
# Do Y
ds= ds.y()
...
# Do Z
ds = ds.z()
```

I think you can write the code as 
```
ds = ds.x() \ # Do X
       .y() \ # D0 Y
       .z()  # D0 Y
```

Reply:
I still think this is better for debugging. Tensorflow dataset is hard to debug. Making dataset this way allows debugger to step through datasets and checkout output.

Reply:
I see. OK, that make sense. Leave it AS IS then.